business/noticeService.js

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Microsoft Corporation and others. Licensed under the MIT license.
+// SPDX-License-Identifier: MIT
+
+const { get } = require('lodash')
+const DocBuilder = require('tiny-attribution-generator/lib/docbuilder').default
+const TextRenderer = require('tiny-attribution-generator/lib/outputs/text').default
+const TemplateRenderer = require('tiny-attribution-generator/lib/outputs/template').default
+const JsonSource = require('tiny-attribution-generator/lib/inputs/json').default
+
+const logger = require('../providers/logging/logger')
+
+class NoticeService {
+  constructor(definitionService, attachmentStore) {
+    this.definitionService = definitionService
+    this.attachmentStore = attachmentStore
+    this.logger = logger()
+  }
+
+  async generate(coordinates, template = '') {
+    const renderer = template ? new TemplateRenderer(template) : new TextRenderer()
+    const docbuilder = new DocBuilder(renderer)
+    const result = await this.definitionService.getAll(coordinates)
+    const packages = await Promise.all(
+      Object.keys(result).map(async id => {
+        const definition = result[id]
+        return {
+          name: [definition.coordinates.namespace, definition.coordinates.name].filter(x => x).join('/'),
+          version: get(definition, 'coordinates.revision'),
+          license: get(definition, 'licensed.declared'),
+          copyrights: get(definition, 'licensed.facets.core.attribution.parties'),
+          website: get(definition, 'described.projectWebsite') || '',
+          text: await this._getPackageText(definition)
+        }
+      })
+    )
+    const source = new JsonSource(JSON.stringify({ packages: packages.filter(x => x.license) }))
+    await docbuilder.read(source)
+    return docbuilder.build()
+  }
+
+  async _getPackageText(definition) {
+    if (!definition.files) return ''
+    const texts = await Promise.all(
+      definition.files
+        .filter(file => file.token && file.natures && file.natures.includes('license'))
+        .map(file => this.attachmentStore.get(file.token))
```

Comment:
might want a throat somewhere in here (or higher) as this will flood the store with requests.

Reply:
I was having that thought too when putting this together. I was actually wondering, should the attachmentStore itself should be the one throttling requests instead of every caller of the store?

The callers don't need to know that it is blob storage vs some db query or something which would have different tolerances.

Reply:
let's put bottleneck in the attachment store set to ~1000


```js
get = limiter.wrap(token => {
   store....
})
```